research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -661,3 +721,24 @@ def shutdown(self) -> None:
         self.shutdown_dataloader(self.val_loader, "val_loader")
         self.shutdown_dataloader(self.test_loader, "test_loader")
         return super().shutdown()
+
+    def update_before_train(self, current_server_round: int) -> None:
+        # Was getting OOM errors that could only be fixed by manually cleaning up RAM
+        # https://github.com/pytorch/pytorch/issues/95462
+        # The above issue seems to be the situation I was in.
+        gc.collect()  # Cleans up unused variables
+        # As the linked issue above points out, calling gc.freeze() greatly reduces the
+        # overhead of garbage collection. (from 1.5s to 0.005s)
+        if current_server_round == 2:
+            # Collect runs even faster if we freeze after the end of the first iteration
+            # Likely because a lot of variables are created in the first pass
+            # If we freeze before the first pass, gc.collect has to check all those
+            # variables
+            gc.freeze()
+
+    def transform_gradients(self, losses: TrainingLosses) -> None:
+        """
+        Apply the gradient clipping performed by the default nnunet trainer. This is
+        the default behaviour for nnunet 2.5.1
+        """
+        torch.nn.utils.clip_grad_norm_(self.model.parameters(), 12)
```

Comment:
Should the max_norm be configurable at all? Since norm has a dependence on the tensor size as well as the scale of the input data

Reply:
I'll add it as a parameter but nnunet uses 12 regardless of the input data or scale. My guess is that 12 is a conservatively large norm and that this is in there just as a precaution